instrumentation-core/spring/spring-webmvc-3.1-core/src/main/java/io/opentelemetry/instrumentation/springwebmvc/WebMVCTracingFilter.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.instrumentation.springwebmvc;
+
+import static io.opentelemetry.instrumentation.springwebmvc.SpringWebMvcDecorator.DECORATE;
+
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.Tracer;
+import java.io.IOException;
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.springframework.core.Ordered;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+public class WebMVCTracingFilter extends OncePerRequestFilter implements Ordered {
+
+  private static final String FILTER_CLASS = "WebMVCTracingFilter";
+  private static final String FILTER_METHOD = "doFilterInteral";
+  private final Tracer tracer;
+
+  public WebMVCTracingFilter(Tracer tracer) {
+    this.tracer = tracer;
+  }
+
+  @Override
+  public void doFilterInternal(
+      HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
+      throws ServletException, IOException {
+    HttpServletRequest req = (HttpServletRequest) request;
+
+    Span serverSpan = DECORATE.startSpan(tracer, req, req, FILTER_METHOD, FILTER_CLASS);
```

Comment:
I am not sure if my thoughts below are correct... @trask @anuraaga what do you think about it?

Among other things there are two objectives in introducing manual instrumentations in this repo.

- To produce roughly the same result as auto-instrumentation does
- To allow auto-instrumentation to reuse code from manual one, if possible.

This filter violates at least the first one. Auto-instrumentation for Spring MVC creates spans with `Span.Kind=INTERNAL`, this filter produces `Span.King=SERVER`.

Reply:
The auto-instrumentation uses bytebuddy advice to add instrumentation to WebApplicationContexts, javax DispatchServerlets, and spring-web HandlerAdapters. I could mirror this approach by using spring aspects to keep things consistent. I thought a cleaner approach would be to use [a spring-web filter](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/filter/OncePerRequestFilter.html) to auto-configure traces. I'm still figuring out the pros and cons for each approach but using filters seems to be the most direct.